File: 01-aop/with-aop/src/main/java/com/example/withaop/service/UserService.java
package com.example.withaop.service;

import com.example.withaop.model.User;
import com.example.withaop.repository.UserRepository;
import org.springframework.stereotype.Service;

/**
 * AOP를 사용한 깔끔한 UserService
 *
 * 장점:
 * 1. 순수한 비즈니스 로직만 포함
 * 2. 공통 관심사는 Aspect로 분리
 * 3. 가독성 향상, 유지보수 용이
 */
@Service
public class UserService {

    private final UserRepository userRepository;

    public UserService(UserRepository userRepository) {
        this.userRepository = userRepository;
    }

    /**
     * 사용자 조회
     *
     * 공통 관심사(성능측정, 로깅, 예외처리, 보안)는
     * Aspect에서 자동으로 처리됩니다!
     */
    public User getUser(Long userId) {
        // 순수한 비즈니스 로직만!
        User user = userRepository.findById(userId);
        System.out.println("  → 사용자 조회: " + user.getName());
        return user;
    }

    /**
     * 사용자 생성
     */
    public User createUser(String name, String email) {
        // 순수한 비즈니스 로직만!
        User user = new User(null, name, email);
        userRepository.save(user);
        System.out.println("  → 사용자 생성: " + user.getName());
        return user;
    }

    /**
     * 사용자 삭제
     */
    public void deleteUser(Long userId) {
        // 순수한 비즈니스 로직만!
        userRepository.deleteById(userId);
        System.out.println("  → 사용자 삭제: " + userId);
    }
}

/*
 * 비교:
 * - Without AOP: 각 메서드 30줄 (비즈니스 로직 2줄 + 공통 관심사 28줄)
 * - With AOP: 각 메서드 5줄 (비즈니스 로직만!)
 *
 * 코드 감소: 83% 감소!
 * 가독성: 월등히 향상!
 * 유지보수: 훨씬 쉬움!
 */

File: 02-di-ioc/4-spring-di/src/main/kotlin/com/example/spring/service/UserService.kt
package com.example.spring.service

import com.example.spring.model.User
import com.example.spring.notification.NotificationService
import com.example.spring.repository.UserRepository
import jakarta.annotation.PostConstruct
import jakarta.annotation.PreDestroy
import org.springframework.stereotype.Service

/**
 * 사용자 서비스 - Spring 버전
 *
 * Level 4 (Spring) 장점:
 * 1. @Service로 자동 등록 및 트랜잭션 지원
 * 2. Constructor Injection 자동 지원 (Kotlin은 @Autowired 불필요)
 * 3. @PostConstruct/@PreDestroy로 Lifecycle 관리
 * 4. @Qualifier로 같은 타입 Bean 선택 가능
 * 5. @Scope로 Bean 범위 설정 가능
 * 6. AOP로 횡단 관심사 처리
 * 7. 프로파일별 설정 (@Profile)
 * 8. 테스트 지원 (@MockBean, @SpyBean)
 */
@Service
class UserService(
    private val userRepository: UserRepository,
    private val notificationService: NotificationService
    // Kotlin에서는 생성자 주입이 기본이므로 @Autowired 불필요
) {
    @PostConstruct
    fun init() {
        println("[UserService] 초기화됨 (@PostConstruct)")
    }

    @PreDestroy
    fun destroy() {
        println("[UserService] 종료됨 (@PreDestroy)")
    }

    fun registerUser(name: String, email: String): User {
        println("\n=== 사용자 등록 시작 ===")

        val user = User(name = name, email = email)
        val saved = userRepository.save(user)
        notificationService.send("환영합니다, ${saved.name}님!")

        println("=== 사용자 등록 완료 ===\n")
        return saved
    }

    fun getUser(id: Long): User? {
        return userRepository.findById(id)
    }

    fun getAllUsers(): List<User> {
        return userRepository.findAll()
    }
}
